### DISCUSSION TITLE
JavaScript Parser

### DISCUSSION THREAD

**Discussion by delfiin** [jotaen/xit#15] [Tooling]
This is good stuff, better than todo.txt. I'm writing a XIT parser for JavaScript.

**Comment by delfiin** [comment]
Is the time arrow. construct supposed to work in this case?

```
[ ] My task ->
    2020
```

**Reply by jotaen** [reply]
No, the [date pattern must immediately follow the arrow prefix `-> `](https://github.com/jotaen/xit/blob/main/Specification.md#due-date), without any other character (such as a newline, or additional spaces) in between.

**Reply by delfiin** [reply]
Thanks for clarifying.


```
[
  { type: 'LINE', line: 1, value: '[ ] My task ->' },
  {
    type: 'WORD',
    value: '[ ]',
    line: 1,
    pos: 1,
    role: 'checkbox',
    status: ' '
  },
  { type: 'WORD', value: 'My', line: 1, pos: 4 },
  { type: 'SPACE', value: ' ', line: 1, pos: 6 },
  { type: 'WORD', value: 'task', line: 1, pos: 7 },
  { type: 'SPACE', value: ' ', line: 1, pos: 11 },
  { type: 'WORD', value: '->', line: 1, pos: 12 },
  { type: 'LINE', line: 2, value: '    2020' },
  { type: 'SPACE', value: '    ', line: 2, pos: 1 },
  { type: 'WORD', value: '2020', line: 2, pos: 5 }
]
```
Tokenizer is working
```
[
  { type: 'LINE', line: 1, value: '[@] Do it -> 2022-04' },
  {
    type: 'WORD',
    value: '[@]',
    line: 1,
    pos: 1,
    role: 'checkbox',
    status: '@'
  },
  { type: 'WORD', value: 'Do', line: 1, pos: 4 },
  { type: 'SPACE', value: ' ', line: 1, pos: 6 },
  { type: 'WORD', value: 'it', line: 1, pos: 7 },
  { type: 'SPACE', value: ' ', line: 1, pos: 9 },
  {
    type: 'WORD',
    value: '-> 2022-04',
    line: 1,
    pos: 10,
    role: 'date',
    date: '2022-04'
  }
]
```

**Comment by delfiin** [comment]
Current status:

- **Tokenizer (100%)**
- **Not-quite-AST, but a tree (100%)**
- Validation (80%)
- Error messages (70%)
- Real AST (60%)

https://github.com/delfiin/xit-parser

After feeding this (https://github.com/jotaen/xit-sublime/blob/main/resources/test.xit) to the parser,
this is what it currently outputs: https://github.com/delfiin/xit-parser/blob/master/test.json

**Reply by jotaen** [reply]
That’s terrific. This is a cool basis for people to write a CLI tool or something else.

In the meantime, I have also published a more polished version of the test data here: https://xit.jotaen.net/syntax-guide
It’s still not 100% complete, but already better than the test file from the Sublime package.

Let me know if (or when) I should link this from [the website](https://xit.jotaen.net/). It might help others to discover it. (Don’t feel pushed, though.)

**Reply by delfiin** [reply]
I'll let you know when it's ready. Your syntax guide clarified some things for me, thanks.
One thing tho, the guide suggests this is a valid `#tag=`. Is it really?

**Reply by jotaen** [reply]
Good catch – the guide is actually not in line with the spec (or rather the current draft version of it). I think this would be [something worth to reconsider](https://github.com/jotaen/xit/issues/27).
